feat(manifest): Upgrade to manifest v3

ERosendo opened the pull request:
This PR upgrades the extension code from Manifest V2 to Manifest V3, the newest version of the Chrome Extensions platform.

Here's a list of the key changes in this PR: 

- Updates in the `manifest.js` file:
   - Change the value of the "manifest_version" field from 2 to 3.

   - Host permissions in Manifest V3 are a separate field. This PR separates the old list into two list `permissions` and `host_permissions`
 
   - Manifest V3 limits exposure by restricting which websites and extensions can access resources( images and fonts) in the extension. Instead of providing a list of files as before, We provide an array of objects, each of which maps a set of resources to a set of URLs or extension IDs.
  
   - Manifest V3 for Chrome has deprecated background pages and scripts in favor of service workers. While Firefox continues to support background pages, Safari offers compatibility with both. To ensure cross-browser functionality, this PR introduces the `service_worker` key within the `background` object while preserving `scripts` key for broader support.
  
   - In the earlier versions, Actions used to be divided into `browser_action` and `page_action`. In Manifest V3, they are unified into a single field named `action`.
  
   - Due to inconsistencies in how Chrome and Firefox handle the `favicon_url` key within the `search_provider` object, this PR temporarily removes this object from the manifest to maintain cross-browser compatibility. A dedicated PR will address the `favicon_url` issue.

- Migrate background script to a single service worker:
   - Manifest V3 introduces support for `ES modules` and the `import` keyword in service worker scripts, enabling improved code modularity and organization. This PR refactors several helper methods from the service worker into a new file, `background.js`, located within the `utils` folder. By isolating helper logic, the service worker is streamlined to focus exclusively on event handling, enhancing code readability and maintainability.
  
   - This PR also addresses https://github.com/freelawproject/recap/issues/237 by refactoring code related to CL API data fetching and browser notifications. We've also removed all instances of the old code and associated helper functions. To improve code organization and efficiency, we've introduced the following changes:
       - Dedicated files: Created background_fetch.js to handle CL API data requests and background_notifier.js for managing browser notifications.
       - Wrapper functions: Implemented dispatchBackgroundFetch in fetch.js and dispatchBackgroundNotifier in notifier.js to serve as entry points for triggering background tasks. 
     
   -  Due to the incompatibility of XMLHttpRequest() with service workers, this PR replaces all instances of XMLHttpRequest() with the fetch() API. This ensures the code functions correctly within the service worker environment.
  
- General code Updates:
   - Manifest V3 introduces the scripting API, replacing the tabs API's executeScript function for script injection. To accommodate this change, this PR updates the manifest file's permissions to include 'scripting'. This is necessary as we've adopted programmatic script injection as outlined in https://github.com/freelawproject/recap-chrome/pull/340. **This change will trigger the permission window when updating the extension.**
  
   - Upgrade to Chrome Runtime API. This PR modernizes the extension codebase by replacing deprecated `chrome.extension` APIs with their supported equivalents in the `chrome.runtime` API.
  
   - Migrating to Promises for Asynchronous Operations. This PR significantly enhances code readability and maintainability by replacing traditional callback functions with modern promise-based asynchronous programming. By leveraging async and await keywords, we've streamlined the process of fetching CL data and triggering notifications. This refactoring aligns with the broader goal of modernizing the RECAP codebase as outlined in https://github.com/freelawproject/recap/issues/370.
  
   - Due to enhanced extension security measures in Manifest V3, injecting custom `script` tags from content scripts is no longer allowed. Previously, this technique was used to tweak **form** submission behavior and extract data from **ACMS**. To address these limitations, this PR implements a new approach.
      
     Content scripts live in an isolated environment, preventing direct access to some data on the webpage. Consequently, we've used the `scripting` API and `executeScript()` method from the service worker to execute custom scripts within the desired context.

      To accommodate this change, the `storeVueDataInSession` and `overwriteFormSubmitMethod` functions have been refactored into wrappers that send messages to the service worker. The core logic has been moved to background script functions named `getAndStoreVueData` and `overwriteSubmitMethod`, respectively. This restructuring enables the execution of custom scripts in a secure and compliant manner.

   - Methods for handling district court, appellate, and ACMS reports have been refactored to utilize a new helper method that returns Promises.

   - Updates the `karma.conf.js` configuration file to align with the project's structural changes. Additionally, necessary mock data has been created to ensure test cases pass successfully
   
   - This PR fixes code formatting by aligning Prettier's `printWidth` setting with ESLint's 80-character limit. Additionally, it enhances code readability by standardizing indentation rules for switch cases and multi-line ternary expressions.

mlissner reviewed (commented):
I looked at a couple files enough to see what was going on, and I made a few little comments. Everything looks pretty good to me. I do wish you hadn't reformatted the code at the same time, but the refactoring patterns make sense otherwise. 

It'll be good to try this out for a bit, so I'm going to start using it as my extension and let's see if we can get a few other techie-types doing so as well. 

We don't usually have others review RECAP PRs, but it's probably time we start doing that. I'll ask Alberto if he can take a look.

Thank you for this epic PR!

mlissner commented:
@albertisfu, I'm sorry if I don't remember correctly, but I think you haven't done too many big reviews on the RECAP extension. Do you think you could take a look at this one and see if everything makes sense and looks good? I took a quick look, but I think we'd all feel better if somebody could spend a little time going over the changes making sure there aren't any mistakes or anything.

Thank you in advance, as always!

mlissner commented:
Random thing: 

 - [ ] Update the readme.

mlissner commented:
Got some errors loading in Firefox:

 - [ ] Resolve errors in FF add-ons page:

    ![image](https://github.com/user-attachments/assets/817cfaab-2916-4a53-bd91-cb78308553f5)

mlissner commented:
Sorry, I keep thinking it'll be the last thing. Let's:

 - [ ] Update to version 2.8. 

I'd like to go to 3.0, but that would require that we get recertified by the gov't, so 2.8 it is.

ERosendo commented on src/content_delegate.js:
```diff
@@ -269,150 +290,178 @@ ContentDelegate.prototype.handleDocketDisplayPage = async function () {
-  }, this);
+  if (!recapLinks.count) return;
+  console.info(
+    'RECAP: Got results from API. Running callback on API results to ' +
```
You're right!

ERosendo commented:
> Got some errors loading in Firefox:

@mlissner You're absolutely right! The warning about the `background.service_worker` property arises because our manifest file includes keys for cross-browser compatibility. We followed the example provided here to update the manifest:

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support

This resource also dives deeper into how different browsers handle these keys, which can be helpful for future reference.

ERosendo commented on src/background.js:
> a bit confusing to have background.js and also utils/background.js.

you're right. I'll rename the file.

albertisfu commented on src/utils/background_fetch.js:
```diff
@@ -0,0 +1,126 @@
+        base: courtListenerURL('dockets'),
+        params: {
+          ...data,
+          source__in: '1,3,5,7,9,11,13,15',
```
I think we need to update the sources here according to the new values added in:
https://github.com/freelawproject/courtlistener/blob/main/cl/search/docket_sources.py

so the docket lookup can be successful in all cases?

albertisfu commented on src/utils/background_fetch.js:
```diff
@@ -0,0 +1,126 @@
+    }
+    console.info(`RECAP: Dispatching ${_options.method} for ${_url}`);
+    // dispatch fetch and return the response
+    fetch(_url, _options)
```
The documentation says that if a `fetch()` request takes more than 5 minutes, the service worker may be terminated. I don't think this is generally an issue, but it could be problematic when uploading large files on slow connections.

In the documentation, they recommend periodically calling an trivial API extension to prevent the service worker from being terminated:
https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers

mlissner replied:
We've had a lot of issues with big files in the past, because upload speeds tend to be slower. We could find some big ones in the DB to test (or IIRC, I think we had one in our QA tests too, maybe?)

ERosendo replied:
@albertisfu I've added the script recommended in the documentation to ensure our background/service script is not terminated prematurely and have modified the extension's fetching methods for compatibility. I used my local environment with the extension pointing to my CL Docker service to test these changes, I can confirm that the extension successfully handles requests that take up to 20 minutes. 

<img width="756" alt="Screenshot 2024-08-26 at 1 57 20 PM" src="https://github.com/user-attachments/assets/6e2cb783-b4ce-4497-b5c6-c989279e7d4e">

albertisfu replied:
Great! Yeah, this is working smoothly.

Would it be a good idea to incorporate the documentation URL for reference within the comment?

https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#keep-sw-alive

ERosendo replied:
> Would it be a good idea to incorporate the documentation URL for reference within the comment?

You're right!. I'll add the URL.

ERosendo replied:
@albertisfu comment updated. Thanks

albertisfu reviewed (commented):
This looks great to me. 
Code refactor made code easier to read in general and it works properly for every function in the extension I tested for District, Appellate and ACMS.

I just left a couple of comments in the code and some extra comments below:

- In this banner for district an appellate docket we need space between `of` and the time the docket was updated.

![Screenshot 2024-08-23 at 10 39 58 a m](https://github.com/user-attachments/assets/a50c92db-4d92-4d06-a024-6ac2f7c7e89b)
<img width="833" alt="Screenshot 2024-08-23 at 5 37 08 p m" src="https://github.com/user-attachments/assets/fc3c4452-4775-4b02-ac8c-e9d4416c0ebf">

- In this banner, we also need some spaces.
RECAP Archive, return a, manner is
![Screenshot 2024-08-23 at 10 47 18 a m](https://github.com/user-attachments/assets/20bbad05-8cea-402f-9a7c-b417a7db9eeb)


- On this appellate page, the "create an alert" link doesn't work:
https://ecf.ca5.uscourts.gov/n/beam/servlet/TransportRoom?servlet=DocketReportFilter.jsp&caseNum=23-10362 

Due to the lack of a `pacer_case_id`, this is also failing in the previous version so I just want to confirm whether it's a bug or if it's because it's not possible to retrieve the `pacer_case_id` on this page.


- Lastly, I wanted to ask if this PR might also cover Safari? I compiled the extension for Safari, but it doesn't work for banners and uploads.

ERosendo commented:
@albertisfu thanks for reviewing this PR. I'll start addressing your comments

> I wanted to ask if this PR might also cover Safari?

No, this PR doesn’t cover Safari. I focused on Edge, Firefox, and Chrome because Safari requires an additional step (running a script). I believe that step would also need adjustments.

albertisfu reviewed (approved):
Thanks! I reviewed the changes, and they worked correctly. 
Just one suggestion to include the documentation URL for the trick to keep the worker alive.
